Source/Scene/ShadowMapShader.js

```diff
@@ -0,0 +1,368 @@
+        for (var i = 0; i < namesLength; ++i) {
+            var name = names[i];
+
+            var sourcesLength = sources.length;
+            for (var j = 0; j < sourcesLength; ++j) {
+                if (sources[j].indexOf(name) !== -1) {
+                    return name;
+                }
+            }
+        }
+
+        return undefined;
+    }
+
+    var normalVaryingNames = ['v_normalEC', 'v_normal'];
+
+    function findNormalVarying(shaderSource) {
+        return findVarying(shaderSource, normalVaryingNames);
+    }
+
+    var positionVaryingNames = ['v_positionEC'];
+
+    function findPositionVarying(shaderSource) {
+        return findVarying(shaderSource, positionVaryingNames);
+    }
+
+    ShadowMapShader.createShadowCastVertexShader = function(vs, isPointLight, isTerrain) {
+        var defines = vs.defines.slice(0);
+        var sources = vs.sources.slice(0);
+
+        if (isTerrain) {
+            defines.push('GENERATE_POSITION');
+        }
+
+        var positionVaryingName = findPositionVarying(vs);
+        var hasPositionVarying = defined(positionVaryingName);
+
+        if (isPointLight && !hasPositionVarying) {
+            var length = sources.length;
+            for (var j = 0; j < length; ++j) {
+                sources[j] = ShaderSource.replaceMain(sources[j], 'czm_shadow_cast_main');
+            }
+
+            var shadowVS =
+                'varying vec3 v_positionEC; \n' +
+                'void main() \n' +
+                '{ \n' +
+                '    czm_shadow_cast_main(); \n' +
+                '    v_positionEC = (czm_inverseProjection * gl_Position).xyz; \n' +
+                '}';
+            sources.push(shadowVS);
+        }
+
+        return new ShaderSource({
+            defines : defines,
+            sources : sources
+        });
+    };
+
+    ShadowMapShader.createShadowCastFragmentShader = function(fs, isPointLight, useDepthTexture, opaque) {
+        var defines = fs.defines.slice(0);
+        var sources = fs.sources.slice(0);
+
+        var positionVaryingName = findPositionVarying(fs);
+        var hasPositionVarying = defined(positionVaryingName);
+        if (!hasPositionVarying) {
+            positionVaryingName = 'v_positionEC';
+        }
+
+        var length = sources.length;
+        for (var i = 0; i < length; ++i) {
+            sources[i] = ShaderSource.replaceMain(sources[i], 'czm_shadow_cast_main');
+        }
+
+        var fsSource = '';
+
+        if (isPointLight) {
+            if (!hasPositionVarying) {
+                fsSource += 'varying vec3 v_positionEC; \n';
+            }
+            fsSource += 'uniform vec4 u_shadowMapLightPositionEC; \n';
+        }
+
+        if (opaque) {
+            fsSource +=
+                'void main() \n' +
+                '{ \n';
+        } else {
+            fsSource +=
+                'void main() \n' +
+                '{ \n' +
+                '    czm_shadow_cast_main(); \n' +
```

Comment:
If you don't already have it on the roadmap, please add using a pass-through fragment shader when we know we can (I want to do the same for picking) since I doubt drivers optimize the shader when we disable color writes and many shaders will not need to discard here.